### PR TITLE
Try matching regex named groups to cms page params.

### PR DESCRIPTION
**Pull request by mjauvin**
if a regex rule contains named groups, groups will get matched to cms page params.

e.g.
```
source path: #/path/(?<slug>.*)#
```
The above named group (slug) will match a ":slug" param in a CMS page url.

### CODE REVIEW

**Comment by jaxwilko** [comment]
Code lgtm, can we add a test case for it?

**Comment by mjauvin** [comment]
> Code lgtm, can we add a test case for it?

I'll add a unit test for this, thanks.

File: classes/RedirectManager.php [hunk 1]
```diff
@@ -328,9 +328,17 @@ private function redirectToCmsPage(RedirectRule $rule): string
...
-        // Strip curly braces from keys
-        foreach ($rule->getPlaceholderMatches() as $placeholder => $value) {
-            $parameters[str_replace(['{', '}'], '', (string) $placeholder)] = $value;
+        if ($rule->isRegexMatchType()) {
+            // Try matching named regex groups to cms page params
+            $pregMatchMatches = $rule->getPregMatchMatches();
+            foreach ($pregMatchMatches as $placeholder => $value) {
+                $parameters[(string) $placeholder] = ltrim($value, '/');
+            }
+        } else if ($rule->isPlaceholdersMatchType()) {
```
Comment (LukeTowers) [inline classes/RedirectManager.php]:
```suggestion
        } elseif ($rule->isPlaceholdersMatchType()) {
```

File: models/Redirect.php [hunk 2]
```diff
@@ -176,9 +176,11 @@ protected static function makeValidator(
                 && $request->get('target_type') === self::TARGET_TYPE_STATIC_PAGE;
         });
 
-        $validator->sometimes('from_url', 'is_regex', static function (Fluent $request): bool {
-            return $request->get('match_type') === self::TYPE_REGEX;
-        });
+        if (method_exists($validator, 'validateIsRegex')) {
```
Comment (LukeTowers) [inline models/Redirect.php]: Why is this required?

Comment (mjauvin) [reply models/Redirect.php]: Because the custom validator added in Plugin.php does not exist when running the tests somehow.

Comment (LukeTowers) [reply models/Redirect.php]: I feel like we should probably fix that issue rather than having a hack in here 😜 Any ideas @jaxwilko?

Comment (mjauvin) [reply models/Redirect.php]: I agree but I spent the better part of the day trying and couldn't figure it out.

Comment (mjauvin) [reply models/Redirect.php]: I removed the "hack". If someone wants to take a shot at it, they're most welcome.

Comment (mjauvin) [reply models/Redirect.php]: Interesting, the custom is_regex validation in tests passes in github but fails in my local environment with this error:
```
$ ./artisan winter:test -p winter.redirect
Running tests for plugin: winter.redirect
PHPUnit 9.5.21 #StandWithUkraine

..........E

Time: 00:09.545, Memory: 46.50 MB

There was 1 error:

1) Winter\Redirect\Tests\Cases\RedirectManagerTest::testTargetCmsPageRedirectWithRegexSource
BadMethodCallException: Method Winter\Storm\Validation\Validator::validateIsRegex does not exist.

/home/w2c/w2c.studioazura.com/current/public/vendor/laravel/framework/src/Illuminate/Validation/Validator.php:1534
/home/w2c/w2c.studioazura.com/current/public/vendor/laravel/framework/src/Illuminate/Validation/Validator.php:609
/home/w2c/w2c.studioazura.com/current/public/vendor/laravel/framework/src/Illuminate/Validation/Validator.php:415
/home/w2c/w2c.studioazura.com/current/public/vendor/winter/storm/src/Database/Traits/Validation.php:308
/home/w2c/w2c.studioazura.com/current/public/vendor/winter/storm/src/Database/Traits/Validation.php:76
/home/w2c/w2c.studioazura.com/current/public/vendor/winter/storm/src/Support/Traits/Emitter.php:182
/home/w2c/w2c.studioazura.com/current/public/vendor/winter/storm/src/Database/Model.php:787
/home/w2c/w2c.studioazura.com/current/public/vendor/winter/storm/src/Database/Model.php:830
/home/w2c/w2c.studioazura.com/current/public/plugins/winter/redirect/tests/cases/RedirectManagerTest.php:455
```

**Comment by LukeTowers** [comment]
@mjauvin @jaxwilko is this ready to be merged?

**Comment by mjauvin** [comment]
yes @LukeTowers

**Comment by LukeTowers** [comment]
@mjauvin is there any documentation that can be added for this, either in the readme or in a help message on the backend form page? An undocumented feature might as well not exist after all 😜

**Comment by mjauvin** [comment]
@LukeTowers there isn't, but I can certainly add some this weekend when I get back.

**Comment by mjauvin** [comment]
I added a note to the DOCUMENTATION.md about the named groups in regex.

**Comment by LukeTowers** [comment]
@mjauvin testing issue is back

**Comment by LukeTowers** [comment]
Shouldn't the plugin's registration of the validation method being moved before the unit test check conditional solve the testing issue?

**Comment by mjauvin** [comment]
@LukeTowers it should but it doesn't. I frankly don't know what's causing this, I spent hours trying to figure it out to no avail

**Comment by mjauvin** [comment]
@LukeTowers moving the registerCustomValidators() method in the Plugin's register() method solved the issue.

**Comment by mjauvin** [comment]
@LukeTowers finally fixed the issue by adding the "is_regex" custom validator within the beforeValidate method of the model.

**Comment by LukeTowers** [comment]
@mjauvin I'm not a huge fan of that, that level of application extension shouldn't be happening dynamically every time validate() is called on a model instance. What problems were you running into with having it registered in the plugin? I feel like having it registered all the time (i.e. bypassing the isTests check) should have fixed the issue, what was the problem you were still running into?

**Comment by mjauvin** [comment]
> @mjauvin I'm not a huge fan of that, that level of application extension shouldn't be happening dynamically every time validate() is called on a model instance. What problems were you running into with having it registered in the plugin? I feel like having it registered all the time (i.e. bypassing the isTests check) should have fixed the issue, what was the problem you were still running into?

@LukeTowers The tests were failing saying the validator does not exist (even when bypassing the isTest()

**Comment by LukeTowers** [comment]
@bennothommo are you able to take a look at this?

**Comment by mjauvin** [comment]
Waiting for https://github.com/wintercms/winter/pull/672 to get merged for tests to pass.

**Comment by mjauvin** [comment]
@LukeTowers https://github.com/wintercms/winter/pull/672 has been merged and tests are now passing.
